### PR TITLE
 - formatting of import long lines

### LINKED ISSUE #4746: imports_granularity does nothing when long import lines are present
On some inputs containing long lines, rustfmt makes no changes, either to those lines or any other imports. Using shorter identifiers fixes the problem.

**To Reproduce**

Run `rustup run nightly rustfmt --verbose --config imports_granularity=Crate --emit stdout` on this input file:

    use foo::bar;
    use foo::{quuz, baz};
    use abaadfsasdfdsfdfas::aasdffjsioejr::abc::sdsdf::sdfsdfsdf::sdfsdfdsf::{
        asdfasdefasdasdfsdfdfasdf::asdfasdasedfafasdfasdf,
    };

The input file will be returned unchanged.

**Expected behavior**

Delete the final `s` from the initial `abaadfsasdfdsfdfas` and the correct behavior is seen; the output is

    use abaadfsasdfdsfdfa::aasdffjsioejr::abc::sdsdf::sdfsdfsdf::sdfsdfdsf::asdfasdefasdasdfsdfdfasdf::asdfasdasedfafasdfasdf;
    use foo::{bar, baz, quuz};

I can reproduce the same behavior with `merge_imports=true` on stable.

**Meta**

- rustfmt version: `rustup run nightly rustfmt --version` returns `rustfmt 1.4.36-nightly (7de6968 2021-02-07)`. 
- From where did you install rustfmt?: `rustup`
- How do you run rustfmt: `rustfmt`

### PR DESCRIPTION
**Pull request by davidBar-On**
Fixes #4746

Suggested fix to format long import lines.  The main issue was that there was no check whether the last import segment in the formatted output will not exceed the maximum width.  In case the import line breaks the next line starts with the original next line indentation.

**No** braces are added when breaking an import line, and it seems that this is not required.

The two test files are the same, except that one is for `imports_granularity` Crate and the other is for Preserve.

The fix changes `issue_3033.rs` test, as the target test file included an import line with length of 106 bytes.  To keep the original target file it is now used as the test source file (there was no source test file for this issue).

### CODE REVIEW

File: src/imports.rs [hunk 1]
```diff
@@ -1077,15 +1077,27 @@ impl Rewrite for UseTree {
     fn rewrite(&self, context: &RewriteContext<'_>, mut shape: Shape) -> Option<String> {
         let mut result = String::with_capacity(256);
         let mut iter = self.path.iter().peekable();
+        let start_shape = shape;
+
         while let Some(segment) = iter.next() {
             let segment_str = segment.rewrite(context, shape)?;
+            let mut added_len = first_line_width(&segment_str);
+            if iter.peek().is_some() {
+                added_len += 2; // 3 == "::"
```
Comment (ytmimi) [inline src/imports.rs:1086]: show this be?
```suggestion
                added_len += 2; // 2 == "::"
```

File: tests/target/issue-4746/crate.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
+// rustfmt-imports_granularity: Crate
+
+// Long imports with reordering - from original issue
+use abaadfsasdfdsfdfas::aasdffjsioejr::abc::sdsdf::sdfsdfsdf::sdfsdfdsf::
+    asdfasdefasdasdfsdfdfasdf::asdfasdasedfafasdfasdf;
```
Comment (ytmimi) [inline tests/target/issue-4746/crate.rs:5]: I don't think this is formatting that's permitted based on the [formatting rules for large imports](https://github.com/rust-lang/fmt-rfcs/blob/master/guide/items.md#large-list-imports). I think we'd need to add braces in order to introduce this line break.

**Comment by ytmimi** [comment]
As mentioned in https://github.com/rust-lang/rustfmt/issues/4746#issuecomment-1152811556 I think these changes need to be version gated.

I can also see cases where users are fine with their imports exceeding the `max_width` and wouldn't want rustfmt to break it over multiple lines. Not sure if we should introduce a new config to opt-in to the new line break behavior being proposed here.

**Comment by davidBar-On** [comment]
> I can also see cases where users are fine with their imports exceeding the `max_width` and wouldn't want rustfmt to break it over multiple lines.

The `UseTree` does not keep the original line breaks or braces when a single import is break between line.  As far as I understand, this information is not available in the AST.  In addition, already now, rustfmt does not keep the original line breaks or braces.

For example, the `UseTree` of the following three imports is the same and currently rusftfmt formats all as `use a::b::c::d::e;`:
```rust
use a::b::c::d::e;
use a::b::c::
    d::e;
use a::b::c::{
    d::e
};
```

File: src/imports.rs (hunk 1 above)
Comment (davidBar-On) [reply src/imports.rs]: Changed (the 3 was because originally I also added `{`)

**Comment by davidBar-On** [comment]
@ytmimi, I improved the formatting, so when an import path is broken between lines, nested imports will always be in a new line, even if they could be in the same line.  This may be more in line of the formatting rules.

I also prepared a private version that always adds braces when import path is broken between lines. It produces the following [Crate](https://github.com/rust-lang/rustfmt/files/10096460/issue-4746-target__crate.rs.txt) and [Preserve](https://github.com/rust-lang/rustfmt/files/10096461/issue-4746-target__preserve.rs.txt) target outputs for the source test cases.  I hope this can help with the decision whether braces should or should not be added whenever the import line is broken between lines.

**Comment by ytmimi** [comment]
> For example, the `UseTree` of the following three imports is the same and currently rusftfmt formats all as use `a::b::c::d::e;`

The first two are the same because the rustc_parser doesn't treat whitespace as significant so both of those get parsed as `a::b::c::d::e;` despite the whitespace. The 3rd example you point out is a little different. it technically gets parsed as `a::b::c::{d::e};`, but we can trivially normaize this to `a::b::c::d::e;`. This might be an oversight when `reorder_imports=false`, but if you set `reorder_imports=false` we don't perform that normalization so you're left with:

```rust
use a::b::c::d::e;
use a::b::c::d::e;
use a::b::c::{d::e};
``` 

I don't expect us to tackle this in this PR. It's unclear whether this is a bug in the `reorder_imports=false` case or if it's a bug in the `reorder_imports=true` case since the default value of [`imports_granularity=Preserve`](https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#imports_granularity) isn't supposed to make changes to your imports.

**Comment by ytmimi** [comment]
> @ytmimi, I improved the formatting, so when an import path is broken between lines, nested imports will always be in a new line, even if they could be in the same line. This may be more in line of the formatting rules.

Just so we're on the same page are you referring to the [`nested import rules`](https://github.com/rust-lang/fmt-rfcs/blob/master/guide/items.md#nested-imports)?

> I also prepared a private version that always adds braces when import path is broken between lines. It produces the following [Crate](https://github.com/rust-lang/rustfmt/files/10096460/issue-4746-target__crate.rs.txt) and [Preserve](https://github.com/rust-lang/rustfmt/files/10096461/issue-4746-target__preserve.rs.txt) target outputs for the source test cases. I hope this can help with the decision whether braces should or should not be added whenever the import line is broken between lines.

Thanks for continuing to work on this. I think including the braces are more inline with the current rules in the style guide, however, I think it might be best to try and get an explicit rule added to the style guide for long imports that can't be formatted on a single line. There are some good edge cases illustrated by your test cases, for example,

```rust
use z123::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::{
    z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::{
        z1234567::z1234567::z1234567::{
            baz, foo
        }
    }
};

// VS

use z123::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::
    z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::
    z1234567::z1234567::z1234567::{baz, foo};
```

@calebcartwright what are your thoughts on this?

**Comment by davidBar-On** [comment]
> Just so we're on the same page are you referring to the [nested import rules](https://github.com/rust-lang/fmt-rfcs/blob/master/guide/items.md#nested-imports)?

Yes, practically... Actually I assumed that from the [Large list imports rules](https://github.com/rust-lang/fmt-rfcs/blob/master/guide/items.md#large-list-imports), but I meant exactly what is written in [Nested imports rules](https://github.com/rust-lang/fmt-rfcs/blob/master/guide/items.md#nested-imports).

> use z123::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::
    z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::
    z1234567::z1234567::z1234567::{baz, foo};

With the last change referred above (when **not** adding braces for each line break) this is now formatted as:
```rust
use z123::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::
    z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::z1234567::
    z1234567::z1234567::{
        baz, foo
    };
``` 

> It's unclear whether this is a bug in the `reorder_imports=false` case or if it's a bug in the `reorder_imports=true` case since the default value of [imports_granularity=Preserve](https://github.com/rust-lang/rustfmt/blob/master/Configurations.md#imports_granularity) isn't supposed to make changes to your imports.

I see that setting `reorder_imports=false` causes several issues in the test cases.  I will try to at least understand what is causing that.

**Comment by ytmimi** [comment]
> With the last change referred above (when not adding braces for each line break) this is now formatted as:

Thanks for clarifying!

> I see that setting reorder_imports=false causes several issues in the test cases. I will try to at least understand what is causing that.

If I had to guess it's because we're not doing any normalization of the imports when `reorder_imports=false` is set as demonstrated above with the `a::b::c::d::e;` example, however without a deeper dive into the code I can't say for certain.

**Comment by davidBar-On** [comment]
> I see that setting `reorder_imports=false` causes several issues in the test cases. I will try to at least understand what is causing that.

Issue was mainly related to indentation. Now fixed.  Added related test cases.

During the issue evaluation, I found that use-trees merge does not take `reorder_imports` config into account.  Submitted a separate PR #5621 to resolve this issue. Therefore, the "use-trees merge" cases were removed from the `crate_no_reorder.rs` test file, since these cases will produce wrong output until PR #5621 (or other fix) will be merged.

> If I had to guess it's because we're not doing any normalization of the imports when `reorder_imports=false` is set as demonstrated above with the `a::b::c::d::e` example, however without a deeper dive into the code I can't say for certain.

See the discussion about the "`a::b::c::d::e` example" issue in the description of PR #5621, including an optional suggested fix.
